pinot-segment-local/src/main/java/org/apache/pinot/segment/local/utils/nativefst/ByteSequenceIterator.java

```diff
@@ -0,0 +1,180 @@
+    this._fst = fst;
+
+    if (fst.getFirstArc(node) != 0) {
+      restartFrom(node);
+    }
+  }
+
+  /**
+   * Restart walking from <code>node</code>. Allows iterator reuse.
+   *
+   * @param node Restart the iterator from <code>node</code>.
+   * @return Returns <code>this</code> for call chaining.
+   */
+  public ByteSequenceIterator restartFrom(int node) {
+    _position = 0;
+    _bufferWrapper.clear();
+    _nextElement = null;
+
+    pushNode(node);
+    return this;
+  }
+
+  /** Returns <code>true</code> if there are still elements in this iterator. */
+  @Override
+  public boolean hasNext() {
+    if (_nextElement == null) {
+      _nextElement = advance();
+    }
+
+    return _nextElement != null;
+  }
+
+  /**
+   * @return Returns a {@link ByteBuffer} with the sequence corresponding to the
+   *         next final state in the automaton.
+   */
+  @Override
+  public ByteBuffer next() {
+    if (_nextElement != null) {
+      final ByteBuffer cache = _nextElement;
+      _nextElement = null;
+      return cache;
+    } else {
+      final ByteBuffer cache = advance();
+      if (cache == null) {
+        throw new NoSuchElementException();
+      }
+      return cache;
+    }
+  }
+
+  /**
+   * Advances to the next available final state.
+   */
+  private final ByteBuffer advance() {
+    if (_position == 0) {
+      return null;
+    }
+
+    while (_position > 0) {
+      final int lastIndex = _position - 1;
+      final int arc = _arcs[lastIndex];
+
+      if (arc == 0) {
+        // Remove the current node from the queue.
+        _position--;
+        continue;
+      }
+
+      // Go to the next arc, but leave it on the stack
+      // so that we keep the recursion depth level accurate.
+      _arcs[lastIndex] = _fst.getNextArc(arc);
+
+      // Expand buffer if needed.
+      final int bufferLength = this._buffer.length;
+      if (lastIndex >= bufferLength) {
+        this._buffer = Arrays.copyOf(_buffer, bufferLength + EXPECTED_MAX_STATES);
+        this._bufferWrapper = ByteBuffer.wrap(_buffer);
+      }
+      _buffer[lastIndex] = _fst.getArcLabel(arc);
+
+      if (!_fst.isArcTerminal(arc)) {
+        // Recursively descend into the arc's node.
+        pushNode(_fst.getEndNode(arc));
+      }
+
+      if (_fst.isArcFinal(arc)) {
+        _bufferWrapper.clear();
+        _bufferWrapper.limit(lastIndex + 1);
+        return _bufferWrapper;
+      }
+    }
+
+    return null;
+  }
+
+  /**
+   * Not implemented in this iterator.
+   */
+  @Override
+  public void remove() {
+    throw new UnsupportedOperationException("Read-only iterator.");
+  }
+
+  /**
+   * Descends to a given node, adds its arcs to the stack to be traversed.
+   */
+  private void pushNode(int node) {
+    // Expand buffers if needed.
+    if (_position == _arcs.length) {
+      _arcs = Arrays.copyOf(_arcs, _arcs.length + EXPECTED_MAX_STATES);
+    }
+
+    _arcs[_position++] = _fst.getFirstArc(node);
+  }
+}
```

Comment:
(nit) new line at the end, same for other files

Reply:
Strange. My IDE shows newlines at the end of all files, and I think Checkstyle checks the same as well?